logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
         return conjunction;
     }
 
-    private void ensureResolvedConformToTheirSuper(Conjunction conjunction) {
-        Set<Variable> visited = new HashSet<>();
-        conjunction.variables().forEach(variable -> ensureResolvedConformToTheirSuper(variable, visited));
-    }
-
-    private TypeVariable above(Variable variable) {
-        if (variable.isType()) {
-            if (variable.asType().sub().isPresent()) return variable.asType().sub().get().type();
-            return null;
-        } else if (variable.isThing()) {
-            if (variable.asThing().isa().isPresent()) return variable.asThing().isa().get().type();
-            return null;
-        } else throw GraknException.of(ILLEGAL_STATE);
-    }
-
-    private void ensureResolvedConformToTheirSuper(Variable variable, Set<Variable> visited) {
-        if (variable == null || visited.contains(variable) || variable.reference().isLabel()) return;
-        visited.add(variable);
-        ensureResolvedConformToTheirSuper(above(variable), visited);
-        removeVarsViolatingSuper(variable);
-    }
-
-    private void removeVarsViolatingSuper(Variable subVar) {
-        TypeVariable supVar = above(subVar);
-        if (supVar == null) return;
-        if (supVar.reference().isLabel()) return;
-
-        Set<Label> supLabels = supVar.resolvedTypes();
-        Set<Label> subLabels = subVar.resolvedTypes();
-        if (supLabels.isEmpty()) return;
-
-        if (subLabels.isEmpty()) {
-            Set<Label> subResolvedOfSupLabels = supLabels.stream().flatMap(label -> getType(label).getSubtypes())
-                    .map(Type::getLabel).collect(Collectors.toSet());
-            subVar.addResolvedTypes(subResolvedOfSupLabels);
-            return;
-        }
+    public Conjunction resolveVariables(Conjunction conjunction) {
+        resolveLabels(conjunction);
```

Comment:
We always call `resolveLabels` because it's always necessary to do that before running `resolveVariables`.